bug/60458 The 'set reminder' action is available to users who are not signed in

akabiru opened the pull request:
# Ticket
<!-- Provide the link to respective work package -->

<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->

https://community.openproject.org/work_packages/60458

# What are you trying to accomplish?
<!-- Provide a description of the changes. -->

Set reminder action should not be visible anonymous users. This PR performs the data correction by removing `manage_own_reminders` permissions from non member and anonymous roles.

## Screenshots
<!-- Provide before/after screenshots, videos, or graphs for any visual changes; otherwise, remove this section -->

| Before | After |
|--------|--------|
| <img width="1581" alt="Screenshot 2025-01-08 at 3 04 51 PM" src="https://github.com/user-attachments/assets/e4163abe-dbb8-4b92-b39b-f97e48f8fc74" /> | <img width="1511" alt="Screenshot 2025-01-08 at 3 05 49 PM" src="https://github.com/user-attachments/assets/08ba9e6f-3d51-4935-8774-9ed91005ee08" /> |


# What approach did you choose and why?
<!-- This section is a place for you to describe your thought process in making these changes.
     List any tradeoffs you made to take on or pay down tech debt.
     Describe any alternative approaches you considered and why you discarded them. -->


17b94f6f55432d105cf5e7129ae1fdcc77d1bf4e introduced a seed migration that added the `manage_own_reminders` permission to roles with `view_project` permission via `Migration::MigrationUtils::PermissionAdder`.

However, the permission added did not correctly exclude non member and anonymous roles. This has been corrected in https://github.com/opf/openproject/pull/17554

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

akabiru committed:
[#60458] The 'set reminder' action is available to users who are not signed in

https://community.openproject.org/work_packages/60458

17b94f6f55432d105cf5e7129ae1fdcc77d1bf4e introduced a seed migration that added the
`manage_own_reminders` permission to roles with `view_project` permission
via `Migration::MigrationUtils::PermissionAdder`.

However, the permission added did not correctly exclude non member and anonymous roles.
This has been corrected in https://github.com/opf/openproject/pull/17554

This PR performs the data correction by removing `manage_own_reminders` permissions from
non member and anonymous roles.

akabiru committed:
Clean out any anonymous user created reminder notifications

akabiru commented on db/migrate/20250108100511_remove_incorrect_manage_own_reminders_permission.rb:
```diff
@@ -0,0 +1,16 @@
+class RemoveIncorrectManageOwnRemindersPermission < ActiveRecord::Migration[7.1]
+  def up
+    # Remove manage_own_reminders permission from non member and anonymous roles
+    RolePermission
+      .where(
+        role: Role.where(builtin: [Role::BUILTIN_NON_MEMBER, Role::BUILTIN_ANONYMOUS]),
+        permission: :manage_own_reminders
+      ).destroy_all
+
+    # Remove all reminders created by anonymous user
+    Reminder.where(creator_id: AnonymousUser.first.id).destroy_all
```
@machisuji I presume low risk in using AR models as these are unlikely to change and break migrations, plus the tests should help I think. Pls shout if you disagree 🍺

machisuji replied:
@akabiru Thanks for writing a test for the migration! And yet I'm afraid I'd still prefer it if you didn't use AR in the migration code. However unlikely the permissions are to change, they could. Or something could change with the Role model, or the Reminder.

This should be 2 fairly simple SQL statements. Could I please ask you to use SQL?

machisuji replied:
P.S. And thanks for flagging me! <3

klaustopher replied:
With the Reminders I could kind of agree, but with the role and permission stuff, I'd personally say that it is pretty set in stone. The test will tell us if we break things, which we will probably neve will

machisuji replied:
Unless the test is running on the schema at the time of writing, not on the latest schema (which it does), it actually won't tell us if the migration breaks.

akabiru replied:
Thanks for the review @machisuji, happy to update to raw SQL and thanks for providing context yet again :)

akabiru committed:
Drop AR in migrations for continuity

In case the models change in the future, then the migrations would start failing.
See: https://github.com/opf/openproject/pull/17556#discussion_r1908488686

akabiru committed:
Add missing LEGAL blurb

machisuji commented on db/migrate/20250108100511_remove_incorrect_manage_own_reminders_permission.rb:
```diff
@@ -0,0 +1,33 @@
+    execute <<-SQL.squish
+      DELETE FROM role_permissions
+      WHERE role_id IN (
+        SELECT id FROM roles WHERE builtin IN (#{Role::BUILTIN_NON_MEMBER}, #{Role::BUILTIN_ANONYMOUS})
```
@akabiru Thanks a lot for turning this into SQL! Looks pretty easy to understand.

About this line in particular, I know it's pretty much guaranteed not to change. I know it's tedious when I nitpick like this. But could you please also get rid of the used constants and just use their values directly (non member => 1 and anonymous => 2)? We might as well go all the way here.

akabiru replied:
@machisuji , I do see a risk in hard-coding these if the enums are ever reassigned for any reason. It would be more straightforward to find all references of these "constants"- which by definition should never change. But if we use the actual values then I worry this can fail silently.

akabiru replied:
Ok, I think I found a sensible middle ground- added references in comments. Not ideal IMO, but should be good enough.

https://github.com/opf/openproject/pull/17556/commits/0e77873f52fd0db2008339305d407991d0a085c0

akabiru committed:
Go all the way with hardcoded values

machisuji reviewed (approved):
👍
